### PR TITLE
Improve the codegen of the vector accelerated `System.Numerics.*` types

### PR DESCRIPTION
**Pull request by tannergooding**
This improves the codegen for more scenarios involving `Vector2/3/4`, `Quaternion`, and `Plane`.

### Note about Quaternion and Plane

In an ideal world, particularly for `Quaternion` and `Plane`, a large portion of the JIT side changes of this PR would be unnecessary. However, we don't live in that world and these types were all exposed with public fields which limits the ability to change their internal implementation details.

What this means is that even though `Quaternion` should be wrapping a single `Vector4` field and deferring most of its operations to `Vector4`, it can't. And unlike `Matrix4x4` where it is sufficiently large and therefore the limitation could be worked around using `Unsafe.As` to convert the `byref`, avoid the copy, and still end up getting promoted; the same is not trivially true for `Quaternion` or `Plane` which are enregisterable.

So what ends up happening is that we explicitly track `Quaternion` and `Plane` as being a `TYP_SIMD16` much as we do for `Vector4`. This allows us to light up all the same methods with all the same intrinsic handling.

I considered limiting things to a couple "bitcast" intrinsics, `AsVector4()` and `AsQuaternion/Plane` , as this would allow implementing `operator +` as `(value1.AsVector4() + value2.AsVector4()).AsQuaternion`. However, given we already have the intrinsic handling for `operator +` for Vector4 directly and this avoids the need to handle 3 intrinsics + inlining + eliding the copies, I opted to do it this way instead.

### Note about other changes

The other APIs are split between intrinsic and managed handling.

For the intrinsic handling, I limited it to APIs which already had intrinsic handling (e.g. `Add` and `operator +` can both be trivially done) and otherwise to APIs which are functionally "primitive" operations for these types such as `Length`. This allows us to avoid inlining, extra intrinsic recognition/matching, and better handle problematic scenarios such as these being `instance methods` (when ideally they'd be `static extensions` like they are for `Vector64/128/256/512<T>`) or returning scalar types just to immediately convert back to a vector.

Other APIs such as `Transform`, `Cross`, `Reflect`, etc are considered "complex" and are left entirely to managed. They do not optimize down to 1-2 instruction sequences and are expected to be more expensive and so are left to managed code.

### CODE REVIEW

**Comment by tannergooding** [comment]
Many diffs are cases such as:
```diff
-; 0 inlinees with PGO data; 4 single block inlinees; 0 inlinees without PGO data
+; 0 inlinees with PGO data; 2 single block inlinees; 0 inlinees without PGO data
```

or
```diff
-;* V04 tmp2         [V04    ] (  0,  0   )   simd8  ->  zero-ref    ld-addr-op "NewObj constructor temp"
-;* V05 tmp3         [V05    ] (  0,  0   )   simd8  ->  zero-ref    ld-addr-op "NewObj constructor temp"
-;* V06 tmp4         [V06    ] (  0,  0   )   simd8  ->  zero-ref    V01.X(offs=0x00) P-INDEP "field V01.X (fldOffset=0x0)"
-;* V07 tmp5         [V07    ] (  0,  0   )   simd8  ->  zero-ref    V01.Y(offs=0x08) P-INDEP "field V01.Y (fldOffset=0x8)"
-;* V08 tmp6         [V08    ] (  0,  0   )   simd8  ->  zero-ref    V01.Z(offs=0x10) P-INDEP "field V01.Z (fldOffset=0x10)"
-;* V09 tmp7         [V09,T01] (  0,  0   )   simd8  ->  zero-ref    V03.X(offs=0x00) P-INDEP "field V03.X (fldOffset=0x0)"
-;* V10 tmp8         [V10,T02] (  0,  0   )   simd8  ->  zero-ref    V03.Y(offs=0x08) P-INDEP "field V03.Y (fldOffset=0x8)"
-;* V11 tmp9         [V11,T03] (  0,  0   )   simd8  ->  zero-ref    V03.Z(offs=0x10) P-INDEP "field V03.Z (fldOffset=0x10)"
+;* V04 tmp2         [V04    ] (  0,  0   )   simd8  ->  zero-ref    V01.X(offs=0x00) P-INDEP "field V01.X (fldOffset=0x0)"
+;* V05 tmp3         [V05    ] (  0,  0   )   simd8  ->  zero-ref    V01.Y(offs=0x08) P-INDEP "field V01.Y (fldOffset=0x8)"
+;* V06 tmp4         [V06    ] (  0,  0   )   simd8  ->  zero-ref    V01.Z(offs=0x10) P-INDEP "field V01.Z (fldOffset=0x10)"
+;* V07 tmp5         [V07,T01] (  0,  0   )   simd8  ->  zero-ref    V03.X(offs=0x00) P-INDEP "field V03.X (fldOffset=0x0)"
+;* V08 tmp6         [V08,T02] (  0,  0   )   simd8  ->  zero-ref    V03.Y(offs=0x08) P-INDEP "field V03.Y (fldOffset=0x8)"
+;* V09 tmp7         [V09,T03] (  0,  0   )   simd8  ->  zero-ref    V03.Z(offs=0x10) P-INDEP "field V03.Z (fldOffset=0x10)"
```

Both of these represent less work for the JIT to be doing.

**Comment by tannergooding** [comment]
Some of the instance methods do show small regressions. These are because the instance method takes the first parameter as a `byref` and so they'll generate slightly different code (most frequently an extra local) to account for this.

In practice, such methods are intrinsic themselves and so this code isn't actually hit and doesn't represent any real regression.

**Comment by tannergooding** [comment]
The change to morph to recognize `Create(ScalarDot(..., ...))` and `Create(ScalarSqrt(ScalarDot(..., ...)))` gives some nice wins in various places:
```diff
-       vdpps    xmm3, xmm2, xmm2, 113
-       vsqrtss  xmm3, xmm3
-       vbroadcastss xmm3, xmm3
+       vdpps    xmm3, xmm2, xmm2, 127
+       vsqrtps  xmm3, xmm3
```

Other functions have "regressions" which are just code size increases due to methods that were supposed to already be getting inlined actually being inlined now thanks to them being intrinsic.

And then of course APIs involving `Quaternion` and `Plane` themselves have improvements ranging from small to large. For example, `Quaterion.get_IsIdentity` was:
```asm
; Assembly listing for method System.Numerics.Quaternion:get_IsIdentity():bool:this
; Emitting BLENDED_CODE for X64 CPU with AVX - Windows
; optimized code
; rsp based frame
; partially interruptible
; No PGO data
; 0 inlinees with PGO data; 2 single block inlinees; 1 inlinees without PGO data
; Final local variable assignments
;
;  V00 this         [V00,T01] (  3,  3   )   byref  ->  rcx         this single-def
;# V01 OutArgs      [V01    ] (  1,  1   )  lclBlk ( 0) [rsp+00H]   "OutgoingArgSpace"
;  V02 tmp1         [V02,T00] (  5, 10   )  struct (16) [rsp+08H]   do-not-enreg[SF] "impAppendStmt"
;* V03 tmp2         [V03    ] (  0,  0   )  struct (16) zero-ref    do-not-enreg[SF] "struct address for call/obj"
;* V04 tmp3         [V04    ] (  0,  0   )  struct (16) zero-ref    ld-addr-op "NewObj constructor temp"
;  V05 tmp4         [V05,T02] (  3,  2   )    bool  ->  rax         "Inline return value spill temp"
;* V06 tmp5         [V06    ] (  0,  0   )  struct (16) zero-ref    "Inlining Arg"
;* V07 tmp6         [V07    ] (  0,  0   )  struct (16) zero-ref    "Inlining Arg"
;* V08 tmp7         [V08,T07] (  0,  0   )   float  ->  zero-ref    V04.X(offs=0x00) P-INDEP "field V04.X (fldOffset=0x0)"
;* V09 tmp8         [V09,T08] (  0,  0   )   float  ->  zero-ref    V04.Y(offs=0x04) P-INDEP "field V04.Y (fldOffset=0x4)"
;* V10 tmp9         [V10,T09] (  0,  0   )   float  ->  zero-ref    V04.Z(offs=0x08) P-INDEP "field V04.Z (fldOffset=0x8)"
;* V11 tmp10        [V11,T10] (  0,  0   )   float  ->  zero-ref    V04.W(offs=0x0c) P-INDEP "field V04.W (fldOffset=0xc)"
;  V12 tmp11        [V12,T03] (  2,  2   )   float  ->  mm0         V06.X(offs=0x00) P-INDEP "field V06.X (fldOffset=0x0)"
;  V13 tmp12        [V13,T04] (  2,  1.50)   float  ->  mm1         V06.Y(offs=0x04) P-INDEP "field V06.Y (fldOffset=0x4)"
;  V14 tmp13        [V14,T05] (  2,  1.50)   float  ->  mm2         V06.Z(offs=0x08) P-INDEP "field V06.Z (fldOffset=0x8)"
;  V15 tmp14        [V15,T06] (  2,  1.50)   float  ->  mm3         V06.W(offs=0x0c) P-INDEP "field V06.W (fldOffset=0xc)"
;* V16 tmp15        [V16    ] (  0,  0   )   float  ->  zero-ref    V07.X(offs=0x00) P-INDEP "field V07.X (fldOffset=0x0)"
;* V17 tmp16        [V17,T11] (  0,  0   )   float  ->  zero-ref    V07.Y(offs=0x04) P-INDEP "field V07.Y (fldOffset=0x4)"
;* V18 tmp17        [V18,T12] (  0,  0   )   float  ->  zero-ref    V07.Z(offs=0x08) P-INDEP "field V07.Z (fldOffset=0x8)"
;* V19 tmp18        [V19,T13] (  0,  0   )   float  ->  zero-ref    V07.W(offs=0x0c) P-INDEP "field V07.W (fldOffset=0xc)"
;
; Lcl frame size = 24

G_M51825_IG01:
       sub      rsp, 24
       vzeroupper 
						;; size=7 bbWeight=1 PerfScore 1.25
G_M51825_IG02:
       vmovups  xmm0, xmmword ptr [rcx]
       vmovups  xmmword ptr [rsp+08H], xmm0
       vmovss   xmm0, dword ptr [rsp+08H]
       vmovss   xmm1, dword ptr [rsp+0CH]
       vmovss   xmm2, dword ptr [rsp+10H]
       vmovss   xmm3, dword ptr [rsp+14H]
       vxorps   xmm4, xmm4
       vucomiss xmm0, xmm4
       jp       SHORT G_M51825_IG05
       jne      SHORT G_M51825_IG05
						;; size=46 bbWeight=1 PerfScore 21.33
G_M51825_IG03:
       vxorps   xmm0, xmm0
       vucomiss xmm1, xmm0
       jp       SHORT G_M51825_IG05
       jne      SHORT G_M51825_IG05
       vxorps   xmm0, xmm0
       vucomiss xmm2, xmm0
       jp       SHORT G_M51825_IG05
       jne      SHORT G_M51825_IG05
       xor      eax, eax
       vucomiss xmm3, dword ptr [reloc @RWD00]
       setnp    al
       jp       SHORT G_M51825_IG04
       sete     al
						;; size=42 bbWeight=0.50 PerfScore 7.96
G_M51825_IG04:
       jmp      SHORT G_M51825_IG06
						;; size=2 bbWeight=0.50 PerfScore 1.00
G_M51825_IG05:
       xor      eax, eax
						;; size=2 bbWeight=0.50 PerfScore 0.12
G_M51825_IG06:
       add      rsp, 24
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
RWD00  	dd	3F800000h		;         1


; Total bytes of code 104, prolog size 7, PerfScore 43.32, instruction count 29, allocated bytes for code 104 (MethodHash=46a9358e) for method System.Numerics.Quaternion:get_IsIdentity():bool:this
; ============================================================
```

But is now:
```asm
; Assembly listing for method System.Numerics.Quaternion:get_IsIdentity():bool:this
; Emitting BLENDED_CODE for X64 CPU with AVX - Windows
; optimized code
; rsp based frame
; partially interruptible
; No PGO data
; Final local variable assignments
;
;  V00 this         [V00,T00] (  3,  3   )   byref  ->  rcx         this single-def
;# V01 OutArgs      [V01    ] (  1,  1   )  lclBlk ( 0) [rsp+00H]   "OutgoingArgSpace"
;
; Lcl frame size = 0

G_M51825_IG01:
       vzeroupper 
						;; size=3 bbWeight=1 PerfScore 1.00
G_M51825_IG02:
       vmovups  xmm0, xmmword ptr [rcx]
       vcmpps   xmm0, xmm0, xmmword ptr [reloc @RWD00], 0
       vmovmskps xrax, xmm0
       cmp      eax, 15
       sete     al
       movzx    rax, al
						;; size=26 bbWeight=1 PerfScore 10.50
G_M51825_IG03:
       ret      
						;; size=1 bbWeight=1 PerfScore 1.00
RWD00  	dq	0000000000000000h, 3F80000000000000h


; Total bytes of code 30, prolog size 3, PerfScore 15.50, instruction count 8, allocated bytes for code 30 (MethodHash=46a9358e) for method System.Numerics.Quaternion:get_IsIdentity():bool:this
; ============================================================
```

**Comment by tannergooding** [comment]
There are still a few APIs that can be optimized further, but they are all on the managed side of things and involve a more complex refactoring of their code. I plan on handling that in a separate PR

File: src/mono/mono/mini/simd-intrinsics.c
Comment (tannergooding) [inline src/mono/mono/mini/simd-intrinsics.c:1982]: Mono doesn't have an easy way to create SIMD constants that I could find.

These are simply:
* `UnitX: 1, 0, 0, 0`
* `UnitY: 0, 1, 0, 0`
* `UnitZ: 0, 0, 1, 0`
* `UnitW: 0, 0, 0, 1`

File: src/mono/mono/mini/simd-intrinsics.c
Comment (tannergooding) [inline src/mono/mono/mini/simd-intrinsics.c:2122]: All of these, except `Lerp`, depend on `Dot` and half also depend on `Sqrt`.

I initially tried to extract the `SN_Dot` handling to a `emit_dot_vector` helper (as there exists a `emit_sum_vector` helper). However, the general logic for `SN_Dot` is not correct for `Vector2/3/4` here and it is out of scope to ensure that is correctly handled.

If someone could give me some pointers on how to check which of `Vector2/3/4` are being processed, I can fix the logic to do the right thing in a follow up PR.

Namely, the current logic is always using `0xF1` which is "dot product all four elements and place result in element 0". However, `Vector2` should be using `0x31` and `Vector3` should be using `0x71`

Likewise, for `Normalize` we should be getting a full `SIMD` result (not a scalar) and so we want `dot` but using `0x3F`, `0x7F`, or `0xFF` respectively.

Comment (tannergooding) [reply src/mono/mono/mini/simd-intrinsics.c:2122]: `Sqrt` not being supported on WASM yet also makes the logic a bit more complicated than would be ideal.

File: src/mono/mono/mini/simd-intrinsics.c
Comment (tannergooding) [inline src/mono/mono/mini/simd-intrinsics.c:1834]: This method group should also be updated to cover `Quaternion` and `Plane`. I don't have enough context into Mono to know if `Plane` is problematic since it is `Vector3, float`. I imagine `Quaternion` just needs to be handled since it has an identical layout to `Vector4`. However, not all the same functions get lightup for `Quaternion` and it is non-obvious how to correctly filter things here.

**Comment by tannergooding** [comment]
CC. @dotnet/jit-contrib 

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=153646&view=ms.vss-build-web.run-extensions-tab)

As indicated, x64 is pure some size wins in MinOpts. FullOpts has some small size regressions (benchmarks, primarily) since key intrinsic methods are now being properly "inlined" (these are all 1-2 instruction methods, so they should be always inlined). There is likewise a small `-0.01%` TP improvement for libraries_tests.pmi` and no changes otherwise.

For x86 we see pure wins across the board (MinOpts, FullOpts, and TP).

For Arm64 we have likewise see similar size wins for MinOpts but correspondingly see `-0.01%` to `-0.03%` improvements to TP.  FullOpts likewise sees `-0.02%` to `-0.04%` improvements to throughput, and similar size regressions for Windows, but sees more considerable size regressions (again benchmarks, but also libraries_tests) for Linux/MacOS.

**Comment by tannergooding** [comment]
The main reason for these regressions is actually because the benchmarks are "naive" and are using constants. Consider for example:
```csharp
[Benchmark]
public Vector3 TransformByQuaternionBenchmark() => Vector3.Transform(VectorTests.Vector3Value, Quaternion.Identity);
```

and the `Vector3.Transform` method:
```csharp
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static Vector3 Transform(Vector3 value, Quaternion rotation)
{
    float x2 = rotation.X + rotation.X;
    float y2 = rotation.Y + rotation.Y;
    float z2 = rotation.Z + rotation.Z;

    float wx2 = rotation.W * x2;
    float wy2 = rotation.W * y2;
    float wz2 = rotation.W * z2;
    float xx2 = rotation.X * x2;
    float xy2 = rotation.X * y2;
    float xz2 = rotation.X * z2;
    float yy2 = rotation.Y * y2;
    float yz2 = rotation.Y * z2;
    float zz2 = rotation.Z * z2;

    return new Vector3(
        value.X * (1.0f - yy2 - zz2) + value.Y * (xy2 - wz2) + value.Z * (xz2 + wy2),
        value.X * (xy2 + wz2) + value.Y * (1.0f - xx2 - zz2) + value.Z * (yz2 - wx2),
        value.X * (xz2 - wy2) + value.Y * (yz2 + wx2) + value.Z * (1.0f - xx2 - yy2)
    );
}
```

Because `Quaternion` was 4 float fields before this PR, it was getting promoted and the relevant `Identity` constant of `new Quaternion(0, 0, 0, 1)` was therefore allowing most of this logic to be eliminated as dead code.

Now that its `TYP_SIMD16` we get enregistered instead and while we have a `CNS_VEC`, we don't have any constant folding support for `TYP_SIMD*` yet. Adding the basic support for scalars is pretty trivial and I can do that in a follow up PR (it's basically just recognizing `GetElement` and `ToScalar` for `CNS_VEC`, which is "simple" since it returns a scalar `T` and not a `Vector###<T>`).

That will get these wins back for the such cases. It won't however keep them when this function is rewritten to be properly accelerated using SIMD. That would require more advanced constant folding for things like `Vector###<T> + Vector###<T>`. This is also doable, but a slightly larger work item and the real world wins are likely not as prominent as the benchmarks are showing.

**Comment by tannergooding** [comment]
Ignoring the "bad benchmarks" (bad because they are being constant folded to a no-op), this shows up as measurable wins for real world scenarios in a similar vein to the Matrix3x2 and Matrix4x4 rewrite (https://github.com/dotnet/runtime/pull/80091) where we saw 8-10x and even higher perf improvements (we saw up to 48x in perf triage)

I'm working on fixing the dotnet/performance tests to more accurately test these bits of code where possible

**Comment by tannergooding** [comment]
Noting that this also effectively represents the "last" of the `System.Numerics.*` related JIT changes. It should give measurable improvements to our 3rd party libraries, such as ImageSharp and others, which are using these types for SIMD acceleration.

Other future changes are of course possible, but with the removal of the legacy SIMD support, the removal of GT_SIMD, and the general move to have these implemented internally as GT_HWINTRINSIC, we don't have any other functions that are "missing" or otherwise "lacking" from the original 2014 SIMD endeavor.

**Review by TIHan — APPROVED** [review]
JIT changes look good to me, glad we are getting the last bits of the old SIMD taken care of.

I understand the regressions are from tests that don't really test much, and you are going to make a follow-up PR to address it.

**Comment by fanyang-mono** [comment]
I am wondering why none of the CI lanes under `runtime` pipeline was triggered?

**Comment by tannergooding** [comment]
Unclear, maybe a CI outage. I'll close/reopen to retrigger CI.

File: src/mono/mono/mini/simd-intrinsics.c
Comment (tannergooding) [inline src/mono/mono/mini/simd-intrinsics.c:1426]: `arg0_type` for this was just wrong before. `arg0` is `Vector128<T>` and we're extracting a `T`

File: src/mono/mono/mini/simd-intrinsics.c
Comment (tannergooding) [inline src/mono/mono/mini/simd-intrinsics.c:1953]: This was moved to be an extension method hanging off `static class Vector` (which is likewise where it would live if we eventually make it public), so I moved the logic up to handle it.

Comment (tannergooding) [reply src/mono/mono/mini/simd-intrinsics.c:1953]: Same for `SN_WithElement` below

**Comment by tannergooding** [comment]
This just needs sign-off on the Mono side before it can be merged.

**Comment by tannergooding** [comment]
Just tacking some of the Mono issues logged related to this PR:
* https://github.com/dotnet/runtime/issues/81482
* https://github.com/dotnet/runtime/issues/81483

As discussed with @fanyang-mono offline. My current plan, after this PR is merged, is to put up a PR which changes `emit_vector_2_3_4` to actually handle `Vector2/3` (today it only handles `Vector4`). I will optimistically be adding Quaternion/Plane support at the same time (#81483) and will update the issue with anything that blocks the latter two (I don't expect any blockers for `Vector2/3`).

File: src/mono/mono/mini/simd-intrinsics.c
Comment (fanyang-mono) [inline src/mono/mono/mini/simd-intrinsics.c:1132]:
```suggestion
	} else {
```

File: src/mono/mono/mini/simd-intrinsics.c
Comment (fanyang-mono) [inline src/mono/mono/mini/simd-intrinsics.c:1134]: I am not sure if we want to assert here, since mono won't disable assertion in release build. I think it is better to return false when the class is not one of those expected ones.

File: src/mono/mono/mini/simd-intrinsics.c
Comment (fanyang-mono) [inline src/mono/mono/mini/simd-intrinsics.c:1127]:
```suggestion
	if (vector_type->type == MONO_TYPE_GENERICINST) {
```

File: src/mono/mono/mini/simd-intrinsics.c
Comment (fanyang-mono) [inline src/mono/mono/mini/simd-intrinsics.c:1439]:
```suggestion
		if (fsig->params [0]->type == MONO_TYPE_GENERICINST) {
```

File: src/mono/mono/mini/simd-intrinsics.c
Comment (fanyang-mono) [inline src/mono/mono/mini/simd-intrinsics.c:1446]:
```suggestion
		} else {
```

File: src/mono/mono/mini/simd-intrinsics.c
Comment (fanyang-mono) [inline src/mono/mono/mini/simd-intrinsics.c:1707]:
```suggestion
		if (fsig->params [0]->type == MONO_TYPE_GENERICINST) {
```

File: src/mono/mono/mini/simd-intrinsics.c
Comment (fanyang-mono) [inline src/mono/mono/mini/simd-intrinsics.c:1714]:
```suggestion
		} else {
```

File: src/mono/mono/mini/simd-intrinsics.c
Comment (fanyang-mono) [reply src/mono/mono/mini/simd-intrinsics.c:1953]: I can see `SN_WithElement` case is not being removed. Based on what you said earlier, they should be gone too?

**Review by fanyang-mono — COMMENTED** [review]
Thanks for taking care of Mono.

File: src/mono/mono/mini/simd-intrinsics.c
Comment (tannergooding) [reply src/mono/mono/mini/simd-intrinsics.c:1134]: 👍, didn't realize Mono asserts were preserved in release code.

**Comment by tannergooding** [comment]
> I can see SN_WithElement case is not being removed. Based on what you said earlier, they should be gone too?

Yes, looks like I didn't add the change to my commit. I'll fix this and the formatting issues after extra-platforms finishes to avoid additional CI churn.

File: src/mono/mono/mini/simd-intrinsics.c
Comment (tannergooding) [reply src/mono/mono/mini/simd-intrinsics.c:1953]: Did this locally, forgot to `git add`. Will be included in the commit pushed after extra-platforms finishes.

File: src/mono/mono/mini/simd-intrinsics.c
Comment (tannergooding) [reply src/mono/mono/mini/simd-intrinsics.c:1127]: Fixed.

**Comment by tannergooding** [comment]
Link to extra platforms failures is here: https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_build/results?buildId=156061

Other than a timeout due to an offline machine, the test failures match what the scheduled runs are hitting: https://dev.azure.com/dnceng-public/public/_build/results?buildId=155434&view=ms.vss-test-web.build-test-results-tab

**Comment by tannergooding** [comment]
Brought in latest main to pickup the CI fix

**Pull request merged**
